Pr overlapping grouping

schoi839 opened the pull request.

Commit:
Implemented sorted insertion grouping.

* Sorted insertion grouping (a type of greedy grouping method)
has been implemented for both qubit wise commuting and fully
commuting groups.
* Added tests to verify that sorted insertion grouping returns
appropriately commutative groups.

Commit:
Implementation of overlapping sorted insertion (coefficient splitting).

* Introduced class OverlappingGroups for initializing the mutually
commuting fragments with overlapping coefficients and for optimizing
these coefficients (according to cov_dict).
* Added tests for the newly added OverlappingGroups.
* Implemented __eq__ in BinaryPauliString and added a corresponding test.

ThomsonYen commented on src/tequila/grouping/binary_rep.py:
```diff
@@ -230,23 +232,41 @@ def anti_commutativity_matrix(self):
         return matrix @ gram @ matrix.T % 2
 
-    def commuting_groups(self, method='rlf'):
+    def commuting_groups(self, method='rlf', condition=None, overlap_help=None):
         """
         Return the partitioning of the hamiltonian into commuting groups.
         List of BinaryHamiltonian's
         """
```
Because it's not so obvious what the parameters are doing, some explanation following the format below is appreciated https://github.com/schoi839/tequila/blob/915605c2001c3f0839898944ad5d131e3436e0dc/src/tequila/circuit/circuit.py#L242-L263

ThomsonYen commented on src/tequila/grouping/binary_rep.py:
```diff
@@ -230,23 +232,41 @@ def anti_commutativity_matrix(self):
-            raise TequilaException(f"There is no algorithm {method}")
-        return [BinaryHamiltonian(value) for key, value in colors.items()]
 
+        if method_class(method) == 'mcc':
+            cg = self.anti_commutativity_matrix()
+            if method == 'lf':
+                colors = largest_first(terms, n, cg)
+            elif method == 'rlf':
+                colors = recursive_largest_first(terms, n, cg)
+            return [BinaryHamiltonian(value) for key, value in colors.items()]
```
Currently `mcc` methods always works as if `condition="fc"`, regardless of the actual value of `condition`. For example, `commuting_groups('rlf', condition="qwc")` still gives FC groups instead of QWC groups. It'd be nice if the code prevents this.

ThomsonYen commented on src/tequila/grouping/binary_rep.py:
```diff
@@ -230,23 +232,41 @@ def anti_commutativity_matrix(self):
+                colors = recursive_largest_first(terms, n, cg)
+            return [BinaryHamiltonian(value) for key, value in colors.items()]
+        elif method_class(method) == 'greedy':
+            if condition == None: condition = 'fc'
```
Can replace this line by setting `condition="fc"` instead of `condition=None` at function definition

ThomsonYen commented on src/tequila/grouping/overlapping_methods.py:
```diff
@@ -0,0 +1,203 @@
+        self.cov_dict = cov_dict
+        self.n_iter = n_iter
+
+class OverlappingGroups_wo_fixed:
```
Let's keep python class in CamelCase, maybe `OverlappingGroupsWoFixed`

ThomsonYen commented on src/tequila/grouping/overlapping_methods.py:
```diff
@@ -0,0 +1,203 @@
+class OverlappingGroups_wo_fixed:
+    '''
+    Class required for mapping OverlappingGroups onto a vector of overlapping coefficients.
+    Same as OverlappingGroups, but Pauli products corresponding to fixed coefficients are removed.
```
Could you explain what fixed coefficients are? If it's difficult making a reference to something specific in Aadi's paper would help.

ThomsonYen commented on src/tequila/grouping/overlapping_methods.py:
```diff
@@ -0,0 +1,203 @@
+            '''
+            Remove fixed coefficients from o_groups and term_exists_in.
+            These fixed coefficients are determined from other Pauli coefficients in order that
+            the condition sum_alpha c^{alpha} = 0 is satisfied.
```
`c^{alpha} = 0` depends on how you define c^{alpha}. It's probably easiest to reference an equation in Aadi's paper

ThomsonYen commented on src/tequila/grouping/overlapping_methods.py:
```diff
@@ -0,0 +1,203 @@
+                fixed_grp.append(term_exists_in[term_idx][-1])
+                o_groups_wo_fixed[fixed_grp[term_idx]].remove(term)
+                term_exists_in_wo_fixed[term_idx].remove(fixed_grp[term_idx])
+            Ncoeff_grp = np.array([len(lst) for lst in o_groups_wo_fixed])
```
Let's keep snake case for variables,`n_coeff_grp`

ThomsonYen commented on src/tequila/grouping/overlapping_methods.py:
```diff
@@ -0,0 +1,203 @@
+        return None
+
+def get_cov(term1, term2, cov_dict):
+    '''
+    Return the covariance between Pauli string term 1 and 2 from a dictionary.
+    '''
+    if (term1.binary_tuple(), term2.binary_tuple()) in cov_dict:
+        return cov_dict[(term1.binary_tuple(), term2.binary_tuple())]
+    if (term2.binary_tuple(), term1.binary_tuple()) in cov_dict:
+        return cov_dict[(term2.binary_tuple(), term1.binary_tuple())]
+
```
This assumes that `term1` and `term2` commute. This assumption should be in docstring.

ThomsonYen commented on tests/test_binary_pauli.py:
```diff
@@ -152,6 +167,95 @@ def test_commuting_groups():
+    Verifies if the number of variable coefficients are correct.
+    '''
+    o_groups = prep_o_groups() 
+    Ncoeff_num = np.sum(o_groups.wo_fixed.Ncoeff_grp)
+    Ncoeff_num_2 = 0
+    for list_of_term_pos in o_groups.wo_fixed.term_exists_in:
+        Ncoeff_num_2 += len(list_of_term_pos)
+    assert Ncoeff_num == Ncoeff_num_2
```
Snake case for variables.

Commit:
Addressed comments by Thomson (tymcr).

kottmanj commented on src/tequila/grouping/binary_rep.py:
```diff
@@ -230,23 +232,54 @@ def anti_commutativity_matrix(self):
+
+        Parameters
+        ----------
+        method: str: The method used to group the commutative terms together. In the case of overlapping methods,
```
would be good to list the possibie keys here: e.g. rlf (recursive largest first), lf (largest first), other (with explanation) .....

kottmanj commented on src/tequila/grouping/binary_rep.py:
```diff
@@ -230,23 +232,54 @@ def anti_commutativity_matrix(self):
+            """
+            if (method == 'lf' or method == 'rlf'): 
+                mc = 'mcc' 
+                if condition != "fc": raise TequilaException(f"lf and rlf can only return fully commuting fragments.")
```
Good for the user is to explain a bit what went wrong:
e.g.add something in the exception message like:
"combination of method={} and condition={} is not valid. E.g. lf and rlf can only return fully commuting fragments, so condotion needs to be condition=fc".format(method,condition)

Same for the other exceptions

kottmanj commented on src/tequila/grouping/binary_rep.py:
```diff
@@ -230,23 +232,54 @@ def anti_commutativity_matrix(self):
+            elif (method == 'si' or method == 'osi'):
+                mc = 'greedy'
+            else:
+                raise TequilaException(f"There is no algorithm {method}")
```
thats good. Maybe even better:
"there is not method={}.format(method)"
then the user knows, that the "method" key is meant

kottmanj commented on src/tequila/grouping/binary_rep.py:
```diff
@@ -230,23 +232,54 @@ def anti_commutativity_matrix(self):
+        elif method_class(method, condition) == 'greedy':
+            if method == 'si': groups = sorted_insertion_grouping(terms, condition)
+            if method == 'osi':
+                if overlap_aux == None: raise TequilaException(f"Overlapping SI grouping requires a dictionary of covariances.")
```
add something like this to exception message:
"....., call with overlap_aux=X where X is the dictionary"
then the user has an idea what to do

Commit:
Passing options from top-level down to commuting_groups.

Commit:
Implementation of suggested sample size.

* If one wishes to evaluate commuting groups using the suggested sample
  size, it can be done by specifying samples="auto-total#ofsamples",
  e.g., samples="auto-1000000" when compiling the objective (or
  circuit).
* Currently, suggested samples require a covariance dictionary.
  However, in the future, automatic evaluation of covariances will be
  implemented.

Commit:
Bugfix: Overlapping did not consider new Paulis.

* Pauli products can only be introduced to a fragment when it commutes
  with all elements. However, the commutativity with the newly
  introduced Paulis were not considered. This commit fixes the bug.
* Added ignore_const in init_from_qubit_hamiltonian such that we do not
  waste computation effort coefficient splitting the constant term.

Commit:
Addressing comments from kottmanj.

kottmanj committed:
Merge pull request #253 from tequilahub/devel

Updating Master

kottmanj committed:
Update setup.py

schoi839 committed:
Merge branch 'tequilahub:master' into pr-overlapping-grouping
